src/poh_recorder.rs

```diff
@@ -14,82 +14,98 @@ use std::sync::{Arc, Mutex};
 pub enum PohRecorderError {
     InvalidCallingObject,
     MaxHeightReached,
+    MinHeightNotReached,
+}
+
+#[derive(Clone)]
+pub struct WorkingLeader {
+    pub bank: Arc<Bank>,
+    pub sender: Sender<Vec<Entry>>,
+    pub min_tick_height: u64,
+    pub max_tick_height: u64,
 }
 
 #[derive(Clone)]
 pub struct PohRecorder {
     poh: Arc<Mutex<Poh>>,
-    max_tick_height: u64,
+    tick_cache: Arc<Mutex<Vec<Entry>>>,
 }
 
 impl PohRecorder {
-    pub fn max_tick_height(&self) -> u64 {
-        self.max_tick_height
-    }
-
-    pub fn hash(&self) -> Result<()> {
+    pub fn hash(&self) {
         // TODO: amortize the cost of this lock by doing the loop in here for
         // some min amount of hashes
         let mut poh = self.poh.lock().unwrap();
 
-        self.check_tick_height(&poh)?;
-
         poh.hash();
+    }
 
+    fn flush_cache(&self, leader: &WorkingLeader) -> Result<()> {
+        let mut cache = vec![];
+        std::mem::swap(&mut cache, &mut self.tick_cache.lock().unwrap());
+        if !cache.is_empty() {
+            for t in &cache {
+                leader.bank.register_tick(&t.id);
+            }
+            leader.sender.send(cache)?;
+        }
         Ok(())
     }
 
-    pub fn tick(&mut self, bank: &Arc<Bank>, sender: &Sender<Vec<Entry>>) -> Result<()> {
+    pub fn tick(&self, leader: &WorkingLeader) -> Result<()> {
         // Register and send the entry out while holding the lock if the max PoH height
         // hasn't been reached.
         // This guarantees PoH order and Entry production and banks LastId queue is the same
         let mut poh = self.poh.lock().unwrap();
 
-        self.check_tick_height(&poh)?;
+        Self::check_tick_height(&poh, leader).map_err(|e| {
+            let tick = Self::generate_tick(&mut poh);
+            self.tick_cache.lock().unwrap().push(tick);
```

Comment:
Just to clarify my understanding:

1) So for a Tpu, this will generate virtual tick until it reaches min_tick_height (the height at which this leader slot starts), and then flush all those virtual ticks to broadcast which will signal the beginning of this node's leader slot. Is this the basis for how timeouts on each node will work? If so, do these ticks need to be reset every time the node votes?

2) How will the Tvu be interacting with this new PoHService?